sdk/src/main/scala/io/horizen/account/sc2sc/AbstractCrossChainMessageProcessor.scala

```diff
@@ -0,0 +1,173 @@
+      throw new ExecutionRevertedException("Reached maximum number of CrosschainMessages per epoch: request is invalid")
+    }
+
+    val request = AccountCrossChainMessage(messageType, sender.toBytes, receiverSidechain, receiver, payload)
+    val messageHash = CryptoLibProvider.sc2scCircuitFunctions.getCrossChainMessageHash(AbstractCrossChainMessageProcessor.buildCrosschainMessageFromAccount(request, networkParams))
+
+    //check for duplicates in this and other message processor, in any epoch
+    if (view.getCrossChainMessageHashEpoch(messageHash).nonEmpty) {
+      throw new ExecutionRevertedException("Dupicate crosschain message")
+    }
+
+    val nextNum: Int = numOfReqs + 1
+    setMessageEpochCounter(view, currentEpochNum, nextNum)
+
+
+    val requestInBytes = request.bytes
+    val messageKey = getMessageKey(currentEpochNum, nextNum)
+    view.updateAccountStorageBytes(contractAddress, messageKey, requestInBytes)
+    //we store also the mapping [message hash, epochnumber]
+    view.updateAccountStorageBytes(contractAddress, messageHash.bytes, Ints.toByteArray(currentEpochNum))
+
+    val event = AddCrossChainMessage(sender, messageType, receiverSidechain, receiver, payload)
+    val evmLog = getEthereumConsensusDataLog(event)
+    view.addLog(evmLog)
+
+    request.encode
+  }
+
+  private[horizen] def getMessageEpochCounter(view: BaseAccountStateView, epochNum: Int) = {
+    val key = getMessageEpochCounterKey(epochNum)
+    val counterInBytesPadded = view.getAccountStorage(contractAddress, key)
+    val counterInBytes = counterInBytesPadded.drop(counterInBytesPadded.length - Ints.BYTES)
+    val num = Ints.fromByteArray(counterInBytes)
+    num
+  }
+
+  private[horizen] def setMessageEpochCounter(view: BaseAccountStateView, currentEpochNum: Int, nextNumOfCrossChainMessages: Int): Unit = {
+    val nextNumOfCrossChainMessagesBytes = Ints.toByteArray(nextNumOfCrossChainMessages)
+    val paddedNextNumOfCrossChainMessages = Bytes.concat(new Array[Byte](32 - nextNumOfCrossChainMessagesBytes.length), nextNumOfCrossChainMessagesBytes)
+    val wrCounterKey = getMessageEpochCounterKey(currentEpochNum)
+    view.updateAccountStorage(contractAddress, wrCounterKey, paddedNextNumOfCrossChainMessages)
+  }
+
+  private[horizen] def calculateKey(keySeed: Array[Byte]): Array[Byte] = {
+    Keccak256.hash(keySeed)
+  }
+
+  private[horizen] def getMessageEpochCounterKey(withdrawalEpoch: Int): Array[Byte] = {
+    calculateKey(Bytes.concat("crossChainMsgEpochCounter".getBytes, Ints.toByteArray(withdrawalEpoch)))
+  }
+
+  private[horizen] def getMessageKey(withdrawalEpoch: Int, counter: Int): Array[Byte] = {
+    calculateKey(Bytes.concat("crossChainMsg".getBytes, Ints.toByteArray(withdrawalEpoch), Ints.toByteArray(counter)))
+  }
+}
+
+object AbstractCrossChainMessageProcessor {
+  private[horizen] def buildCrosschainMessageFromAccount(data: AccountCrossChainMessage, params: NetworkParams): CrossChainMessage = {
```

Comment:
same as for UTXO model: let's move this method to `AccountCrossChainMessage` class.